### ISSUE TITLE
Align JS GC memory manager to current state

### ISSUE THREAD

**Issue by dmik** [bitwiseworks/mozilla-os2#255]
As the `testGCAllocator` failure shows here https://github.com/bitwiseworks/mozilla-os2/issues/254#issuecomment-358609794, the code allocating memory for garbage collector management is not up to date: memory it allocates isn't ordered as expected on OS/2. This may have various side effects including OOMs (out of memory conditions sometimes seen during extensive use of JS).

**Comment by dmik** [comment]
Given that Win and OS/2 memory APIs are very similar, it really makes sense to follow the Win code path closely instead of inventing our own. This code (as well as the `testGCAllocator` test) needs a fixed `DosAllocMemEx` which is currently broken in kLIBC. I fixed it in my test build already, testing it all together now.

**Comment by dmik** [comment]
Okay, `testGCAllocator` works well now with fixed `DosAllocMemEx` and fresh Win code slightly adapted to OS/2. Now it's time to test Firefox itself. This requires a heavy rebuild as some make files were changed at some earlier stage.

**Comment by dmik** [comment]
Built and checked Firefox — all works (the JS slowness is still there). The `DosAllocMemEx` fix is there: http://trac.netlabs.org/libc/ticket/384. The patched DLL is here: http://rpm.netlabs.org/test/libc066.dll. The new XUL.DLL with the GC allocator fix to test for OOMs is here: http://rpm.netlabs.org/test/x2.7z.

**Comment by LewisR** [comment]
I seem to be having trouble getting this new xul.dll to load (error 87). I updated libc066.dll from the link above, downloaded the new xul.dll, and tested. firefox.exe will not load:
`Couldn't load xul.dll (OS/2 error 87).`
Checking xul.dll with PMDLL says that everything should load. libcx0.dll is:

` 1-08-18  14:05          38,644    124   ___A_  libcx0.dll`

Am I missing something else?

**Comment by dmik** [comment]
@LewisR Well, I guess you rebooted after you updated LIBC066.DLL, didn't you? There shouldn't be anything else, the rest is taken from RPM versions. Anyway, it looks like XUL.DLL is still using the old LIBC DLL.

**Comment by dspiatkowski** [comment]
Lewis & Dimitriy,

Here is what I did:

1) repmod -u on the libc066.dll & replace with the test release
2) lxlite xul.dll with '/yua /ydd /yxd /ynl /anp /b- /cs+ /d /i- /ml1 /mr2  
/mf2 /r+ /u+ /x- /zs:0 /zx /zd'
3) re-boot

Firefox came up fine, no problems at the moment, but it has only been  
literally 10-15 mins at most. I usually do not start seeing issues unless I  
do heavy browsing (multiple tabs, as always...lol).

-Dariusz

**Comment by LewisR** [comment]
I did reboot, yes. As long as I know I didn't miss anything else, I'll give this another try. Thanks for confirming.

Thanks for the extra info, Dariusz. I also tried stripping out the debug stuff (though my report was against the dll as it came out of the 7zip, so as not to introduce any other variables). Test system is a ThinkPad T520 running ArcaOS 5.0.1.x beta (after 5.0.1.3 and before 5.0.1.6, with later driver fixes applied).

**Comment by StevenLevine** [comment]
@LewisR, did you redirect stdout/stderr to capture the error?  Was there anything else in the captured info?  FWIW, DosLoadModule usually fails with something other than ERROR_INVALID_PARAMETER (87).

If this problem persists, I can show you how to use OS/2 Trace to possibly see exactly why the API is failing.

Did you check for .trp and popuplog entries, just in case.

**Comment by LewisR** [comment]
No need. I figured it out.

I imported a Foomatic PPD into PSPRINT.DRV, which wreaked havoc with the native print subsystem. Previous FF builds did not try to initialize printing upon program load, but apparently, 45.9 does. This results in a nice SYS3175 in POPUPLOG.OS2 pointing to PSPRINT.DRV. When I backed out all of the FF changes (original xul.dll, libcx - from the last updated xul.dll - , libc) and still got the same error, I went looking for more breakage. Meanwhile, 38.8 and SM 42 continue to work with no issue.
